### ISSUE TITLE
Why not publish apks assets of releases on Github?

### ISSUE THREAD

**Issue by corneliusroemer** [corona-warn-app/cwa-app-android#514]
Yes, you've mentioned that you don't want to support F-Droid.

But what about putting the apks for all releases on Github as assets?

Many of the critical issues arising today could have been avoided, had you simply published the apks in advance - allowing early adopters to check for problems in devices you consider too exotic to be tested by your team.

If an apk can appear on https://apkpure.com/de/corona-warn-app/de.rki.coronawarnapp why can't it appear here on Github? It's more trustworthy, it would allow testing of bug fixes and features before official release and also avoid the country store problems like #478 

What are your reasons against this? Is it not literally just the upload of a file?

**Comment by Loxad** [comment]
Afaik Play Services are needed for the contact tracing, so side loading may be unwanted in general.

**Comment by corneliusroemer** [comment]
@Loxad You may know more than me. Are play services not independent of the apk? Is there any difference between Play store apk and sideload apk? I thought they were 100% identical if they're simply copied. Including signature etc

Just an alternative way of distributing. Or maybe I misused sideload. Feel free to explain.

**Comment by tomjschwanke** [comment]
> 
> 
> @Loxad You may know more than me. Are play services not independent of the apk? Is there any difference between Play store apk and sideload apk? I thought they were 100% identical if they're simply copied. Including signature etc
> 
> Just an alternative way of distributing. Or maybe I misused sideload. Feel free to explain.

The API used by the App is only available with Google Play Services installed. If you have them installed, you also have the Playstore to download the app. 

#477

**Comment by corneliusroemer** [comment]
> > @Loxad You may know more than me. Are play services not independent of the apk? Is there any difference between Play store apk and sideload apk? I thought they were 100% identical if they're simply copied. Including signature etc
> > Just an alternative way of distributing. Or maybe I misused sideload. Feel free to explain.
> 
> The API used by the App is only available with Google Play Services installed. If you have them installed, you also have the Playstore to download the app.
> 
> #477

@tomjschwanke Nope sorry, you're wrong. The app is geo restricted in the playstore.

**Comment by tomjschwanke** [comment]
> @tomjschwanke Nope sorry, you're wrong. The app is geo restricted in the playstore.
Thats been confirmed and they're working on lifting that, see #478

**Comment by corneliusroemer** [comment]
> > @tomjschwanke Nope sorry, you're wrong. The app is geo restricted in the playstore.
> 
> You're right, it's not available everywhere, however they've acknowlegded this and are working on lifting that, see #478

Yes, correct. But why not do both. Plus there's the advance beta testing advantage before it gets pushed to the masses if the apk is available here on Github.

**Comment by tomjschwanke** [comment]
> Yes, correct. But why not do both. Plus there's the advance beta testing advantage before it gets pushed to the masses if the apk is available here on Github.

I must admin, a release on GitHub would be a nice bonus to download old versions (if you want that) and you wouldn't have to go through Google Play. 
Other stores would be out of scope though, since they'd need to upload them to every single store.

But Google Play + GitHub release would be optimal actually

**Comment by vmx** [comment]
> The API used by the App is only available with Google Play Services installed. If you have them installed, you also have the Playstore to download the app.

You can have Google Play Services installed without having a Google Account connected to it. This way you could use the app, but you cannot install it over the Playstore. I think having an official APK download would be way better having those folks downloading if from random third parties.

**Comment by jbauerrfid** [comment]
A few words of clarification: GitHub is a SCM (Source Code Management System). It is not intended to host compiled binaries. Another issue might be that APKs must be signed with publishing keys to put them on Google Play, and IMO the RKI does not want to expose these keys to avoid issues with hacked or fake apps, or malware injection.
If you need the APK you still have the option to clone this git repo and compile the APK yourself with Android Studio from https://developer.android.com/studio

**Comment by cbiller** [comment]
I vote for a already precompiled apk as well. It is all about trust. I trust the sourcecode, but I don´t trust Google, Apple oder the Government - isn´t that the reason why the APP is Opensource? I know I can compile it, but thats time consuming and not all the people have the Know-How to do it. With siedloading I am the one who controls the updates - not Google.

**Comment by tomjschwanke** [comment]
> I vote for a already precompiled apk to download as well. It is all about trust. I trust the sourcecode, but I don´t trust Google, Apple oder the Government - isn´t that the reason why the APP is Opensource? I know I can compile it, but thats time consuming and not all the people have the Know-How to do it. With siedloading I am the one who controls the updates - not Google.

The entire Exposure Notification API is from Google. This app interfaces with it.

**Comment by Pltiton** [comment]
I know, but thats no reason not to sideload it.

**Comment by corneliusroemer** [comment]
> A few words of clarification: GitHub is a SCM (Source Code Management System). It is not intended to host compiled binaries. Another issue might be that APKs must be signed with publishing keys to put them on Google Play, and IMO the RKI does not want to expose these keys to avoid issues with hacked or fake apps, or malware injection.
> If you need the APK you still have the option to clone this git repo and compile the APK yourself with Android Studio from https://developer.android.com/studio
> 

@jbauerrfid Nope this doesn't work, since you need to have a whitelisted key to access the exposure API

**Comment by iMartyn** [comment]
> A few words of clarification: GitHub is a SCM (Source Code Management System). It is not intended to host compiled binaries.

That's simply not true.  Git is an SCM, GitHub is much more than that, we're currently talking in  a thread in the Issue Tracking feature of GitHub and there is a Releases feature for the very request we are talking about.

> APKs must be signed with publishing keys to put them on Google Play, IMO the RKI does not want to expose these keys to avoid issues with hacked or fake apps, or malware injection.

That's not how public-private-key encryption works, signing of an APK is done with a private key that is not in the APK at all, it is verified by using the public key.

Compiled APKs on the release page would go a long way to building trust.  There are valid reasons to have the google services installed and not want to use the play store, and the lack of APKs (that are already compiled as part of CI so can as easily be uploaded to Github as the Play Store) will lead people to getting it via third parties where you don't control the platform, and then malware etc. is a problem.

You're never going to stop APKs getting out there into the wild, by providing them here, you will make this the default place for people who want them, and save people from infections (of both kinds!).

One last point - by not providing the APKs here, it makes it look as if you're trying to hide something, to stop people decompiling the APKs to check that they are actually built from this source.  This undermines the benefits of releasing this as opensource.

**Comment by nilsalex** [comment]
Could someone please clarify this point for me: An officially signed APK (probably extracted from the app bundle? -- I'm not really familiar with all this) would have access to ENF API, just like the version installed from the play store?

**Comment by iMartyn** [comment]
The app bundle is the APK in android world terms, but yes, as far as I know there is no way to lock down API usage to applications installed in one way or another.  There are ways for applications to interrogate the play store API to prevent usage if they were not "purchased" but no API lockdown per se.

**Comment by corneliusroemer** [comment]
Dear maintainers (@tkowark @SebastianWolf-SAP @jakobmoellersap)

What's the status on this? Would be great if you could comment since there is absolutely nothing the community itself can do without you.

While you are working out the play store issue, isn't this a limited but good workaround?

**Comment by immibis** [comment]
@jbauerrfid GitHub has the "releases" feature, which is designed for hosting compiled binaries.

**Comment by nilsalex** [comment]
@jbauerrfid The thing is, you cannot compile the app yourself. Well, you can, but without using the official key to sign it, the app is pretty much useless because it cannot access the API.

**Comment by corneliusroemer** [comment]
This issue is very related to the unfortunately closed issue #477

**Comment by iMartyn** [comment]
@corneliusroemer There has been NO explanation as to why you will not provide an APK for those who have google play services but do not wish to install via an APK, nor for those who wish to verify the built APK is the result of the source.  By closing this issue with no valid reasoning, again, this retracts from the the trust of the application and endangers users.  please reconsider.

**Comment by corneliusroemer** [comment]
@iMartyn I agree with your points. But I'm confused why I'm tagged and why you posted what you wrote here :)

**Comment by vmx** [comment]
For everyone who wants to install the app without the Play Store, it's a security risk as you need to download it from random websites like:

https://apktada.com/app/de.rki.coronawarnapp
https://www.apkmirror.com/apk/robert-koch-institut/corona-warn-app/corona-warn-app-1-0-0-release/corona-warn-app-1-0-0-android-apk-download/

I downloaded version 1.0.0 from my phone (I got it via Aurora Store) and from the websites above. The downloads have the same checksums as the file I got from the phone. They are:

```
MD5: eee459f2b1533a39fbac76e4ded254c9
SHA1: 4f2fe3fd93f2f538153acdbe304b27880443af3c
SHA256: a2c7979dd32f05cc1bd93d992a382f9b60c8556641e34458588bfbee65d927b2
Blake2: 165f3224e3497fef75371b3182d1d1bc5cc581a250550a57a20c57f95077dd0483432fc1c2af7cab9b3c3dd6223ad1acdb8d6e6d2332b36246bf08b3209bf105
```

Even if you cannot provide the APK, it would be nice if you could provide official checksums. It would make APK downloads from website at least a bit safer and you wouldn't need to trust strangers posting checksums.

**Comment by iMartyn** [comment]
@corneliusroemer sorry about the confusion, on the mobile interface the "mentioned a closed issue" and "closed this issue" line look almost identical, I thought that you had closed this issue! :-D

**Comment by Matombo** [comment]
Well this issue is what prevents me from installing the app, too.
Not getting the app outside of the playstore is activly blocking people from using the app, which is against the intended purpose i guess?

**Comment by Pltiton** [comment]
@vmx : Thank you for the Information, wasn´t the most recent App version the 1.02? The one in the store is 1.0.0

**Comment by tomjschwanke** [comment]
> @vmx : Thank you for the Information, wasn´t the most recent App version the 1.02? The one in the store is 1.0.0

Yesterday an update to 1.0.2 was published to the Google Playstore

**Comment by corneliusroemer** [comment]
> @corneliusroemer sorry about the confusion, on the mobile interface the "mentioned a closed issue" and "closed this issue" line look almost identical, I thought that you had closed this issue! :-D
@iMartyn All good, thanks for clarifying. If you read a bit around on my comments on this project you will notice that I'm in fact doing the opposite of what you thought I did. I reopen issues that are closed without proper justification and get myself into hot water with the maintainers. See #478 and #600.

**Comment by nilsalex** [comment]
If the agreement with the Bundesregierung is to only publish via the stores (maybe this is the "legal issue"?), surely they could re-negotiate this point? Would go a long way towards building trust in the app if it could be used without a google account and solve the problems related to availability.

**Comment by SecJoe** [comment]
You have to accept their decision if they don't want to provide APK assets in GitHub. Nevertheless, you can extract the APK from you Android (lets say an old testing device or an emulator) if you have downloaded it before with Google Play. Afterwards, install it on your real device via sideload. The APK is locatated in the internal app folder and you can get it without root. So that's an option to solve your problems, but accept if they want to upload every APK here as a service. Anyway, with the view to OpenSource, APK asset uploads on lets say GitHub are best practice.

**Comment by Pltiton** [comment]
The Government is spending a hell of money to generate trust, goes the Open Source way, to generate trust. I do trust the programmers - but I don´t trust the government. I have automatic updates turned on, and I don´t want to change that just for one app. In the future I could possibly get a complete different code through an update which then generates data profiles - as the Government is the owner of the Google Account, that is easy. The government wants me to install the app - not the other way around. So no apk - no instal - its as simple as that.

**Comment by bersbersbers** [comment]
> I have automatic updates turned on, and I don´t want to change that just for one app.

You don't need to: simply exclude this one app from auto-updates. You're welcome.

> The app will update automatically when updates are available. To turn off automatic updates, uncheck the box.
https://support.google.com/googleplay/answer/113412?hl=en

> Die App wird automatisch aktualisiert, sobald Updates verfügbar sind. Entfernen Sie das Häkchen, wenn keine automatischen Aktualisierungen erfolgen sollen.
https://support.google.com/googleplay/answer/113412?hl=de

**Comment by iMartyn** [comment]
You say 
> You have to accept their decision if they don't want to provide APK assets in GitHub. 
I disagree fundementally - I don't have to accept the decision, and this project is not being open about why they do not want to.
> (snipped random workaround that ignores all the reasons why people want this) So that's an option to solve your problems, but accept if they want to upload every APK here as a service. 
A. This is a bad workaround and shouldn't be necessary and B. They already have a pipeline for building and uploading APKs.  It's not us asking for "a service", it's asking for them to use Github releases as they are intended.
> Anyway, with the view to OpenSource, APK asset uploads on lets say GitHub are used often.
Yes, so at least we agree here, this is the common, expected thing.  When people say "we'll do this open-source on github" it's not unreasonable for people to expect that they do it properly, like most other opensource developers do.

**Comment by nilsalex** [comment]
@iMartyn +1
Although I personally don't care, it should be possible for people who don't want to use a google account to just download an official release of the app and use it on their phone. Especially because there has been no argument to the contrary, just an "not part of the agreement". Any workaround involves using a google account or trusting an unofficial source.

**Comment by ad2003** [comment]
The Problem with this whole project is: It is only partly OpenSource. While I appreciate the idea of making the development an open source project, it all makes absolutely no sense in case of open source, when there is no way to use the app without the closed sources from Google / Apple. 

It is a simply a lie, when you praise this project as Open Source, while you cannot compile this app on your own from GitHub and use it. If you cannot compile the working app from Source, it is not 100% OpenSource. How would someone know what is uploaded on the Store? 

Wouldn't SAP / Telekom / Governemnt have to trust third party (aka Google, Apple) with UserData when using their closed sources? Please correct me if I am wrong here.

**Comment by nilsalex** [comment]
I agree, but this is not in scope for this issue. Here, we are asking why the apk is not being provided. This is a *solvable* issue.

**Comment by ad2003** [comment]
I guess, that is the problem: It is not the idea to have the APKs in the wild, but on the App Stores.

**Comment by iMartyn** [comment]
Again, for clarity, the APKs will always be in the wild from untrusted sources.  There is no feasable way to stop that.  They will also be a massive target for hijacking.  What we are asking for is that they are provided here as a reputable source, to protect users.

**Comment by ad2003** [comment]
No need to "again" me :).

I understand what this topic is about and what is asked here for - apks from reputable source. Yeah, would have been great to have those APKs from day 1 instead from App Stores. Let's see if this going to happen.

**Comment by lordyavin** [comment]
Hey maintainers (@tkowark @SebastianWolf-SAP @jakobmoellersap), please respond to this request. 

The work on this project is or will be paid by an obscenely high amount of tax money. According to [heise](https://heise.de/-4781235 ) **20 million euros** for development and up to **3.5 million** for operation. So there is no reason to ignore justified and meaningful requests from citizens or users.

**Comment by SebastianWolf-SAP** [comment]
Dear colleagues,

similar to the discussion around the [release of the Corona-Warn-App in international app stores](https://github.com/corona-warn-app/cwa-backlog/issues/9) this is something that needs to be discussed with our partners, especially the Robert Koch-Institute. 

As we don't have any update if that can be done (current status [see FAQ](https://www.coronawarn.app/en/faq/#apk)), we will move it to cwa-wishlist for now. Please continue discussing and especially voting, so that things can change...

Mit freundlichen Grüßen/Best regards, 
SW
Corona Warn-App Open Source Team
